akka-stream-tests/src/test/scala/akka/stream/scaladsl/AttributesSpec.scala

```diff
@@ -3,70 +3,238 @@
+
+      // most outer
+      attributes.get[Name] should contain(Name("new-name"))
+
+      // most inner (or defined first)
+      attributes.getFirst[Name] should contain(Name("new-name"))
+      // FIXME the original got replaced, it was not even seen by the materializer when debugging this
+      // might be solved by always wrapping and never replacing attributes
+    }
 
-    "give access to first attribute" in {
-      attributes.getFirst[Name] should ===(Some(Attributes.Name("a")))
+    // just to document the behavior, this creates a nested source with attributes
+    // so they are not really replaced on the inner graph
+    "wrap the attributes on a graph stage" in {
+      val attributes =
+        Source.fromGraph(new AttributesSource(Attributes.name("original-name")))
+          .withAttributes(Attributes.name("nested-source"))
+          .runWith(Sink.head)
+          .futureValue
+
+      // most outer
+      attributes.get[Name] should contain(Name("nested-source"))
+
+      // most inner (or defined first)
+      attributes.getFirst[Name] should contain(Name("original-name"))
     }
 
-    "give access to attribute byt type" in {
-      attributes.get[Name] should ===(Some(Attributes.Name("b")))
+    "use the initial attributes for dispatcher" in {
+      val dispatcher =
+        Source.fromGraph(new ThreadNameSnitchingStage("my-dispatcher"))
+          .runWith(Sink.head)(materializerNoDispatcher)
+          .futureValue
+
+      dispatcher should startWith("AttributesSpec-my-dispatcher")
+    }
+
+    "use the most outer dispatcher when specified directly around the graph stage" in {
+      val dispatcher =
+        Source.fromGraph(
+          // directly on stage
+          new ThreadNameSnitchingStage("akka.stream.default-blocking-io-dispatcher")
+            .addAttributes(ActorAttributes.dispatcher("my-dispatcher")))
+          .runWith(Sink.head)(materializerNoDispatcher)
+          .futureValue
+
+      dispatcher should startWith("AttributesSpec-my-dispatcher")
+    }
+
+    "use the most outer dispatcher when specified further out from the stage" in {
+      val dispatcher =
+        // on the source returned from graph
+        Source.fromGraph(
+          new ThreadNameSnitchingStage("akka.stream.default-blocking-io-dispatcher"))
+          .addAttributes(ActorAttributes.dispatcher("my-dispatcher"))
+          .runWith(Sink.head)(materializerNoDispatcher)
+          .futureValue
+
+      dispatcher should startWith("AttributesSpec-my-dispatcher")
+    }
+
+    "use the dispatcher set in materializer even if dispatcher specified in initial attributes" in {
+      val dispatcher =
+        Source.fromGraph(new ThreadNameSnitchingStage("my-dispatcher"))
+          .runWith(Sink.head)(materializerNoDispatcher)
+          .futureValue
+
+      dispatcher should startWith("AttributesSpec-my-dispatcher")
+    }
+
+    "dispatcher in materializer trumps dispatcher specified directly around the graph stage" in {
```

Comment:
This I disagree with, the materializer dispatcher should only be set if there was no other dispatcher attribute set.

Reply:
+1